### PR TITLE
fix #14873 properly by skipping `abi` field in importc type

### LINKED ISSUE #14873: Codegen bug with lock - no member "abi"
Reopen #12855 under my name to avoid spamming the original poster as Github keeps notifying an issue author even if they unsubscribe.

> Bad C code is generated with the following Nim code:
> ### Example
> 
> ```nim
> import locks
> 
> var l: Lock
> echo l
> ```
> 
> ### Current Output
> 
> ```
> CC: ltest.nim
> Error: execution of an external compiler program 'gcc -c  -w  -I/home/tay/.choosenim/toolchains/nim-1.0.4/lib -I/home/tay -o /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c.o /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c' failed with exit code: 1
> 
> /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c: In function ‘dollar___3Dq1hcGH7a2iNlnh4hnPbg’:
> /home/tay/.cache/nim/ltest_d/stdlib_dollars.nim.c:219:50: error: ‘pthread_mutex_t’ {aka ‘union <anonymous>’} has no member named ‘abi’
>   219 |  addQuoted__N56rmvBL9a3xqnKq09cKdb3A((&result), x.abi);
>       |                                                  ^
> ```
> 
> ### Expected Output
> 
> Compiler either compiles properly or errors before C code gen
> ### Possible Solution
> 
>     * It looks like there might be an issue with the implementation of `$` for the Lock type
> 
> 
> ### Additional Information
> 
>     * Fedora 31: `Linux localhost.localdomain 5.3.11-300.fc31.x86_64 #1 SMP Tue Nov 12 19:08:07 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`
> 
> 
> ```
> $ nim -v
> Nim Compiler Version 1.0.4 [Linux: amd64]
> Compiled at 2019-11-27
> Copyright (c) 2006-2019 by Andreas Rumpf
> 
> git hash: c8998c498f5e2a0874846eb31309e1d1630faca6
> active boot switches: -d:release
> ```

### PR DESCRIPTION
**Pull request by timotheecour**
fix #14873

I can't reproduce https://github.com/nim-lang/Nim/issues/14873 locally, can someone that can reproduces those issues please check that it fails before this PR and works after this PR?

## links:
* https://github.com/nim-lang/Nim/issues/14873#issuecomment-784241605
* https://github.com/nim-lang/Nim/issues/14873#issuecomment-832870351

/cc @shirleyquirk @xflywind @tdely

## note
if this works, we should probably do the same with other apis, eg:
```
lib/posix/posix_linux_amd64.nim:153:5:    abi: array[56 div sizeof(clong), clong]
lib/posix/posix_linux_amd64.nim:157:5:    abi: array[32 div sizeof(clong), clong]
lib/posix/posix_linux_amd64.nim:160:5:    abi: array[4 div sizeof(cint), cint]
lib/posix/posix_linux_amd64.nim:164:5:    abi: array[48 div sizeof(clonglong), clonglong]
lib/posix/posix_linux_amd64.nim:167:5:    abi: array[4 div sizeof(cint), cint]
lib/posix/posix_linux_amd64.nim:171:5:    abi: array[48 div sizeof(clong), clong]
lib/posix/posix_linux_amd64.nim:174:5:    abi: array[4 div sizeof(cint), cint]
lib/posix/posix_linux_amd64.nim:178:5:    abi: array[56 div sizeof(clong), clong]
lib/posix/posix_linux_amd64.nim:181:5:    abi: array[8 div sizeof(clong), clong]
lib/posix/posix_linux_amd64.nim:204:5:    abi: array[32 div sizeof(clong), clong]
lib/posix/posix_linux_amd64.nim:289:5:    abi: array[1024 div (8 * sizeof(culong)), culong]
lib/posix/posix_linux_amd64.nim:298:5:    abi: array[12, int]
lib/posix/posix_linux_amd64.nim:353:5:    abi: array[1024 div (8 * sizeof(clong)), clong]
lib/posix/posix_nintendoswitch.nim:132:5:    abi: array[56 div sizeof(clong), clong]
lib/posix/posix_nintendoswitch.nim:136:5:    abi: array[32 div sizeof(clong), clong]
lib/posix/posix_nintendoswitch.nim:139:5:    abi: array[4 div sizeof(cint), cint]
lib/posix/posix_nintendoswitch.nim:143:5:    abi: array[48 div sizeof(clonglong), clonglong]
lib/posix/posix_nintendoswitch.nim:146:5:    abi: array[4 div sizeof(cint), cint]
lib/posix/posix_nintendoswitch.nim:150:5:    abi: array[48 div sizeof(clong), clong]
lib/posix/posix_nintendoswitch.nim:153:5:    abi: array[4 div sizeof(cint), cint]
lib/posix/posix_nintendoswitch.nim:157:5:    abi: array[56 div sizeof(clong), clong]
lib/posix/posix_nintendoswitch.nim:160:5:    abi: array[8 div sizeof(clong), clong]
lib/posix/posix_nintendoswitch.nim:183:5:    abi: array[32 div sizeof(clong), clong]
lib/posix/posix_nintendoswitch.nim:310:5:    abi: array[((64+(sizeof(clong) * 8)-1) div (sizeof(clong) * 8)), clong]
lib/system/ansi_c.nim:37:9:        abi: array[200 div sizeof(clong), clong]
lib/system/threadlocalstorage.nim:135:9:        abi: array[56 div sizeof(clong), clong]
lib/system/threadlocalstorage.nim:194:8:       abi: array[1024 div (8 * sizeof(culong)), culong]
```

### CODE REVIEW

File: lib/system/syslocks.nim [hunk 1]
```diff
@@ -102,26 +102,18 @@ else:
...
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
-      when defined(linux) and defined(amd64):
-        abi: array[48 div sizeof(clonglong), clonglong]
 
     SysCondAttr {.importc: "pthread_condattr_t", pure, final
                header: """#include <sys/types.h>
                           #include <pthread.h>""".} = object
-      when defined(linux) and defined(amd64):
-        abi: array[4 div sizeof(cint), cint]  # actually a cint
```
Comment (Araq) [inline lib/system/syslocks.nim:124]: Changes like these break `nlvm` (which is still our best idea of how to get good debugging support btw).

Comment (timotheecour) [reply lib/system/syslocks.nim:124]: but why single out this type? other types have no guarantee to have all their fields listed in the importc type (unless `{.completeStruct.}` is given), nor being in order

furthermore, nlvm is heavily behind nim devel: last nim commit that was integrated is Jul 29, 2020, https://github.com/arnetheduck/Nim/commit/bf320ed172f74f60fd274338e82bdc9ce3520dd9)

and there are other ways to do this without resorting to platform specific conditional fields. 

abi/padding fields that aren't used just shouldn't be exposed, which leads to bugs like #14873 etc

## more robust ways to query for fields for an importc type
we could query the type using libclang / cling (https://github.com/timotheecour/Nim/issues/705), leaving guesswork out of the equation. This could also be done optionally, leaving out uncertain fields by default (as done in this PR)

this approach is actually useful for many problems, including all the importc let variables that are currently hardcoded but could instead be queried (see also https://github.com/nim-lang/RFCs/issues/205). It's not hard to do at least on posix.

Comment (alaviss) [reply lib/system/syslocks.nim:124]: Don't we have a `size` pragma? Can it be used as a replacement for `abi`?

Comment (timotheecour) [reply lib/system/syslocks.nim:124]: yes, that would be better than `abi` field, but it still causes the same problems because the ABI for those is not specified and varies by OS/distribution.

note that you can still use `sizeof` at RT; `{.sizeof: N.}` only makes sense if you need it at CT (unlikely because, well, VM is single threaded).

Note also that on osx i get:
```nim
import std/locks
var a: Lock
echo a.sizeof # 64
```
even though no nim abi field is there in the definition of `SysLockObj`, which shows how arbitrary this was. Furthermore, it's also wrong for other architectures besides osx (eg: linux and not amd64)

Long story short: removing the `abi`, as i did in this PR, makes sense.

## links
* https://stackoverflow.com/a/30585442/1426932
> You'll want to use it as their defined type, pthread_mutex_t of course -- since this type will vary by OS / distribution / etc.

Comment (timotheecour) [reply lib/system/syslocks.nim:124]: @araq also you can always do this if you need to access raw bytes:

```nim
import std/[sugar,locks]

proc getBytes[T](buf: var seq[uint8], a: T) =
  let n = T.sizeof
  buf.setLen n
  copyMem(buf[0].addr, a.unsafeAddr, n)

var a: Lock
initLock(a)

echo seq[uint8].default.dup(getBytes(a)) # @[90, 84, 85, 77, 0, 0, 0, 0, ... 85, 77, 90, 84, 85, 77]
```

unlike `abi` field which is fragile and not portable (eg doesn't work on osx etc), this will always work

=> https://github.com/timotheecour/Nim/issues/722

**Comment by Araq** [comment]
Your arguments were convincing.

File: lib/system/syslocks.nim (hunk 1 above)
Comment (Araq) [reply lib/system/syslocks.nim:124]: The abi field never was about accessing bytes, it was about getting sizeof and alignof right.

Comment (timotheecour) [reply lib/system/syslocks.nim:124]: ok; but sizeof and alignof don't need abi fields so long sizeof and alignof are only accessed at RT:
```nim
# after this PR is merged (and on osx) I'm getting:
import std/locks
var a: Lock
echo a.sizeof # 64
echo a.alignof # 8
```

but I think we're agreeing at this point

**Comment by arnetheduck** [comment]
The ABI fields were added because they make certain code work that previously didn't, particularily such code that is careful about memory allocation - this PR breaks such code, and should ideally be rolled back.

File: lib/system/syslocks.nim (hunk 1 above)
Comment (arnetheduck) [reply lib/system/syslocks.nim:124]: > need it at CT

you need it for pointer arithmetic, memory allocation, offset arithmetic and a number of other cases - this PR breaks existing code and introduces differences between compile time and runtime that lead to bugs.

The proper fix is to provide the Nim compiler with correct information at compile time - this can indeed be through libclang - or by just spelling it out.

Comment (arnetheduck) [reply lib/system/syslocks.nim:124]: > nlvm is heavily behind nim devel

nlvm is behind mainly because breaking things and moving fast doesn't work for code that _actually_ depends on the standard library to not break their code, ie actual users of the language and the library - it's PR's like this that prevent it from being updated, as well as drive users of it to look for alternative solutions, including forks - the bar of quality must be higher than "works for me, don't care about the rest" - as long as this is not understood (as can be seen by similar PR's recently), it's a hard sell to actually use the code and the end result is a cycle of "I'll make it work for my case now" style back-and-forth PRs. It's not a sustainable approach.
